chatkit-core/src/test/kotlin/com/pusher/chatkit/users/UserSubscriptionDispatcherTest.kt

```diff
@@ -0,0 +1,193 @@
+package com.pusher.chatkit.users
+
+import com.pusher.chatkit.rooms.api.JoinedRoomApiType
+import com.pusher.chatkit.rooms.api.JoinedRoomApiTypeMapper
+import com.pusher.chatkit.rooms.api.RoomMembershipApiType
+import com.pusher.chatkit.rooms.api.RoomReadStateApiType
+import com.pusher.chatkit.rooms.state.JoinedRoomsState
+import com.pusher.chatkit.rooms.state.roomTwo
+import com.pusher.chatkit.rooms.state.roomTwoId
+import com.pusher.chatkit.state.ChatkitState
+import com.pusher.chatkit.state.JoinedRoom
+import com.pusher.chatkit.state.JoinedRoomsReceived
+import com.pusher.chatkit.state.LeftRoom
+import com.pusher.chatkit.state.RoomDeleted
+import com.pusher.chatkit.state.RoomUpdated
+import com.pusher.chatkit.users.api.UserApiType
+import com.pusher.chatkit.users.api.UserSubscriptionDispatcher
+import com.pusher.chatkit.users.api.UserSubscriptionEvent
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import org.reduxkotlin.Dispatcher
+import org.reduxkotlin.GetState
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+
+class UserSubscriptionDispatcherTest : Spek({
+
+    val simpleJoinedRoomApiType = JoinedRoomApiType(
+            id = "id1",
+            name = "room1",
+            createdById = "person1",
+            pushNotificationTitleOverride = "notification override",
+            private = false,
+            customData = mapOf("highlight" to "blue"),
+            lastMessageAt = "2020-02-27T17:12:10Z",
+            updatedAt = "2020-02-27T17:12:20Z",
+            createdAt = "2020-02-27T17:12:30Z",
+            deletedAt = null
+    )
+
+    val simpleUser = UserApiType(
+            id = "user1",
+            createdAt = "",
+            updatedAt = "",
+            name = "name",
+            avatarURL = null,
+            customData = null,
+            online = false)
+
+    describe("given a user subscription dispatcher with no previous joined room state") {
+
+        val state = mockk<GetState<ChatkitState>>(relaxed = true)
```

Comment:
why do we need `relaxed` here?

Reply:
as per the docs:
> A relaxed mock is the mock that returns some simple value for all functions. This allows to skip specifying behavior for each case, while still allowing to stub things you need. For reference types, chained mocks are returned.

This means i didn't need to mock out any of the methods in our mocks and could just use them.

Reply:
I've looked at removing the word - and the tests don't pass because we'd need to mock more of the dispatcher and state methods. I personally think our current solution where it default mocks things we aren't really interested in but can override the bits we do makes our tests cleaner to follow.

Reply:
what else need to be mocked? essentially GetState is one fun and so is Dispatcher. I have checked that earlier and for me the test was still passing after removing `relaxed`. I tried to check it again but now it does not compile, I guess because of a merge resolution issue?

Reply:
I successfully removed partial and internal-detail mocking [here](https://github.com/pusher/chatkit-android/commit/38599c6651bd890832171ca142f6a438104af7fa#diff-ae4b04dd619cee6ff385e54b0666e2e0L51). Partial mocking is a code smell. It think only for the `Dispatcher` using `relaxed` is OK as essentially it's not our type and we're not concerned about its returned value but just the value we pass to the invocation which is what we verify in the test. Proper mocking of `Dispatcher` would be clumsy and boilerplaty.